v2/pkg/virtualbmc/bmc_server.go

```diff
@@ -0,0 +1,70 @@
+package virtualbmc
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"net"
+
+	"github.com/cybozu-go/log"
+)
+
+// BMCServer represents IPMI Server
+type BMCServer struct {
+}
+
+// Machine defines the interface to manipulate Machine
+type Machine interface {
+	IsRunning() bool
+	PowerOn() error
+	PowerOff() error
+}
+
+// NewBMCServer creates an BMCServer
+func NewBMCServer() (*BMCServer, error) {
+	return &BMCServer{}, nil
+}
+
+func (s *BMCServer) listen(ctx context.Context, addr string, port int, machine Machine) error {
+	serverAddr, err := net.ResolveUDPAddr("udp", fmt.Sprintf("%s:%d", addr, port))
+	if err != nil {
+		return err
+	}
+
+	server, err := net.ListenUDP("udp", serverAddr)
+	if err != nil {
+		return err
+	}
+	go func() {
+		<-ctx.Done()
+		server.Close()
+	}()
+
+	session := NewRMCPPlusSessionHolder()
+	bmcUser := NewBMCUserHolder()
+	bmcUser.AddBMCUser("cybozu", "cybozu")
+
+	buf := make([]byte, 1024)
+	for {
+		_, addr, err := server.ReadFromUDP(buf)
```

Comment:
Is it possible to use a more generic `ReadFrom` paired with `WriteTo`?

Reply:
Do you mean like this?


Server
```go
func Listen(ctx context.Context, conn net.PacketConn, machine Machine) error {
	go func() {
		<-ctx.Done()
		conn.Close()
	}()

	session := NewRMCPPlusSessionHolder()
	bmcUser := NewBMCUserHolder()
	bmcUser.AddBMCUser("cybozu", "cybozu")

	buf := make([]byte, 1024)
	for {
		_, addr, err := conn.ReadFrom(buf)
		if err != nil {
			return err
		}

		bytebuf := bytes.NewBuffer(buf)
		res, err := HandleRMCPRequest(bytebuf, machine, session, bmcUser)
		if err != nil {
			log.Warn("failed to handle RMCP request", map[string]interface{}{
				log.FnError: err,
			})
			continue
		}
		_, err = conn.WriteTo(res, addr)
		if err != nil {
			log.Warn("failed to write to UDP", map[string]interface{}{
				log.FnError: err,
			})
			continue
		}
	}
}
```

Client
```go
var _ = Describe("Virtual BMC", func() {
	It("should turn on and off Machine power via IPMI v2.0", func() {
		serverAddr, err := net.ResolveUDPAddr("udp", fmt.Sprintf("%s:%d", "127.0.0.1", 9623))
		Expect(err).NotTo(HaveOccurred())
		conn, err := net.ListenUDP("udp", serverAddr)
		Expect(err).NotTo(HaveOccurred())
		well.Go(func(ctx context.Context) error {
			return Listen(ctx, conn, &VMMock{running: false})
		})
```

Reply:
Right.